File: ch6/clip4/boj1707/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    static List<Integer> graph[];
    static int color[];

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int k = sc.nextInt();
        for (int tc = 0; tc < k; tc++) {
            int v = sc.nextInt();
            int e = sc.nextInt();
            graph = new List[v + 1];
            color = new int[v + 1];
            for (int i = 0; i <= v; i++) {
                graph[i] = new ArrayList<>();
            }
            for (int i = 0; i < e; i++) {
                int a = sc.nextInt();
                int b = sc.nextInt();
                graph[a].add(b);
                graph[b].add(a);
            }
            for (int i = 1; i <= v; i++) {
                dfs(i, 0);
            }
            boolean isBinary = true;
            for (int i = 1; i <= v; i++) {
                for(int next : graph[i]) {
                    // 인접한 정점끼리 색이 같으면 이분 그래프가 아니다.
                    if (color[i] == color[next]) {
                        isBinary = false;
                        break;
                    }
                }
            }
            if (isBinary) System.out.println("YES");
            else System.out.println("NO");
        }
    }

    static int RED = 1;
    static int BLUE = 2;

    static void dfs(int nodeNum, int depth) {
        if (color[nodeNum] != 0) return;
        if (depth % 2 == 0) color[nodeNum] = RED;
        else color[nodeNum] = BLUE;

        for(int next : graph[nodeNum]) {
            dfs(next, depth + 1);
        }
    }
}

File: ch6/clip4/boj1707/main.py
import sys
sys.setrecursionlimit(10**6)

def dfs(node_num, depth):
    if color[node_num] != 0:
        return
    
    if depth % 2 == 0:
        color[node_num] = RED
    else:
        color[node_num] = BLUE
    
    for next_node in graph[node_num]:
        dfs(next_node, depth + 1)

RED = 1
BLUE = 2

k = int(input())

for _ in range(k):
    v, e = map(int, sys.stdin.readline().split())
    
    graph = [[] for _ in range(v + 1)]
    color = [0] * (v + 1)
    
    for _ in range(e):
        a, b = map(int, sys.stdin.readline().split())
        graph[a].append(b)
        graph[b].append(a)
    
    for i in range(1, v + 1):
        dfs(i, 0)
    
    is_binary = True
    
    for i in range(1, v + 1):
        for next_node in graph[i]:
            if color[i] == color[next_node]:
                is_binary = False
                break
    
    if is_binary:
        print("YES")
    else:
        print("NO")
